File: 프로그래머스/lv3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
from collections import deque
import heapq

def solution(jobs):
    # 지금 시점에 시작할 수 있는 작업이 있다면:
    #   지금 시점에 시작할 수 있는 작업 중 가장 빨리 끝나는 것 수행
    # else:
    #   가장 빨리 시작할 수 있는 작업을 수행

    # job = [도착시간, 소요시간]
    length = len(jobs)
    jobs = deque(sorted(jobs, key=lambda x: (x[0], x[1])))
    heap, time, answer = [], 0, 0
    
    while jobs:
        # 수행 가능한 작업을 heap에 넣는다.
        while jobs and jobs[0][0] <= time:
            s, d = jobs.popleft()
            heapq.heappush(heap, (d, s))
        
        if heap:
            # 수행할 수 있는 작업이 있다면
            # 소요시간이 가장 짧은 것을 수행
            d, s = heapq.heappop(heap)
            answer += (time - s) + d
            time += d
        else:
            # 수행할 수 있는 작업이 없다면
            # 가장 빨리 수행할 수 있는 작업을 수행
            if jobs:
                s, d = jobs.popleft()
                answer += d
                time = s+d
    
    while heap:
        d, s = heapq.heappop(heap)
        answer += (time - s) + d
        time += d
        
    return answer // length

File: 프로그래머스/lv3/42627. 디스크 컨트롤러/디스크 컨트롤러.js
// 현재 시간을 기준으로 가장 빨리 끝나는 작업을 먼저 처리한다.
class Heap {
    constructor(comparator) {
        this.heap = [];
        this.comparator = comparator;
    }

    getLeftChildIndex(parentIndex) {
        return 2 * parentIndex + 1;
    }

    getRightChildIndex(parentIndex) {
        return 2 * parentIndex + 2;
    }

    getParentIndex(childIndex) {
        return Math.floor((childIndex - 1) / 2);
    }

    hasLeftChild(index) {
        return this.getLeftChildIndex(index) < this.heap.length;
    }

    hasRightChild(index) {
        return this.getRightChildIndex(index) < this.heap.length;
    }

    hasParent(index) {
        return this.getParentIndex(index) >= 0;
    }

    leftChild(index) {
        return this.heap[this.getLeftChildIndex(index)];
    }

    rightChild(index) {
        return this.heap[this.getRightChildIndex(index)];
    }

    parent(index) {
        return this.heap[this.getParentIndex(index)];
    }

    swap(index1, index2) {
        const temp = this.heap[index1];
        this.heap[index1] = this.heap[index2];
        this.heap[index2] = temp;
    }

    peek() {
        if (this.heap.length === 0) {
            return null;
        }
        return this.heap[0];
    }

    poll() {
        if (this.heap.length === 0) {
            return null;
        }
        if (this.heap.length === 1) {
            return this.heap.pop();
        }
        const item = this.heap[0];
        this.heap[0] = this.heap.pop();
        this.heapifyDown();
        return item;
    }

    add(item) {
        this.heap.push(item);
        this.heapifyUp();
    }

    heapifyUp() {
        let index = this.heap.length - 1;
        while (this.hasParent(index) && this.comparator(this.parent(index), this.heap[index]) > 0) {
            const parentIndex = this.getParentIndex(index);
            this.swap(parentIndex, index);
            index = parentIndex;
        }
    }

    heapifyDown() {
        let index = 0;
        while (this.hasLeftChild(index)) {
            let smallerChildIndex = this.getLeftChildIndex(index);
            if (
                this.hasRightChild(index) &&
                this.comparator(this.rightChild(index), this.leftChild(index)) < 0
            ) {
                smallerChildIndex = this.getRightChildIndex(index);
            }

            if (this.comparator(this.heap[index], this.heap[smallerChildIndex]) <= 0) {
                break;
            } else {
                this.swap(index, smallerChildIndex);
            }

            index = smallerChildIndex;
        }
    }
    size() {
        return this.heap.length;
    }
}

function solution(jobs) {
    jobs.sort((a, b) => {
        if (a[0] < b[0]) return -1;
        else if (a[0] == b[0]) return 0;
        else return 1;
    });
    const heap = new Heap((a, b) => {
        if (a[1] < b[1]) return -1;
        else if (a[1] === b[1]) return 0;
        else return 1;
    });
    let answer = 0;
    let time = 0;
    let cursor = 0;
    while (cursor < jobs.length || heap.size()) {
        while (cursor < jobs.length) {
            if (jobs[cursor][0] <= time) {
                heap.add(jobs[cursor])
                cursor += 1;
            } else {
                break
            }
        }
        
        if (heap.size()) {
            // 실행할 수 있는 작업 중 가장 빨리 끝나는 작업을 먼저 실행합니다.
            const job = heap.poll();
            answer += time - job[0] + job[1];
            time += job[1];
        } else {
            time += 1
        }
    }

    return Math.floor(answer / jobs.length);
}